Service instance deletion shouldn't be blocked by bindings if the instance was never provisioned

witlessbird opened the pull request:
Closes #1588

n3wscott commented:
ReadyFalse is not enough checking to assume the binding can be ignored. You need to also look at UnbindStatus for the binding to be sure the broker does not need to be notified of the bindings removal.

witlessbird commented:
@n3wscott: could you clarify please? This only deals with instance removal, I assumed that binding will be (eventually?) cleaned up by the binding controller once it detects that the service instance is gone...

n3wscott commented:
@witlessbird I do not believe an instance can be deleted until the bindings are deleted.

witlessbird commented:
@n3wscott: I took a long look at binding reconciliation (here: https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_binding.go#L178). 

It looks to me that binding in `ServiceBindingConditionReady` with `ConditionFalse` or `ServiceBindingConditionFailed` with `ConditionTrue` doesn't actually exist: there may or may not attempts to reconcile errors, but no active binding. The only exception to this I can think of is when a binding existed, but an update to it failed -- I'll take a look at that.

Additionally, my understanding is that should the service instance associated with a binding disappear, orphan mitigation will detect this and remove the binding.

kibbles-n-bytes commented:
@witlessbird: Thanks for the PR!

I think filtering on the Ready condition being False is not quite enough, as @n3wscott said. The Ready condition can be False at any point before a binding is successfully created, but what we'd really like to know is whether a request has been sent to the broker yet for the binding. If it hasn't, then no big deal; we can delete the instance, and the binding will be updated to say it has no associated instance during the next reconcile.

If the request has been sent, however, then we should not allow the associated instance to be deleted, as bindings depend on their associated instance for all broker interactions (see: [here](https://github.com/kubernetes-incubator/service-catalog/blob/c1c69cf27b6a6864366580aa7cde55e9777590e4/pkg/controller/controller_binding.go#L215) and [here](https://github.com/kubernetes-incubator/service-catalog/blob/c1c69cf27b6a6864366580aa7cde55e9777590e4/pkg/controller/controller_binding.go#L732)) and thus we'll be blocked from being able to clean up the binding on the broker side. Orphan mitigation here unfortunately won't help, because it relies on this broker communication as well.

Instead of checking the conditions here, we should check whether `binding.Status.UnbindStatus == v1beta1.ServiceBindingUnbindStatusRequired`, and block instance deletion only if it is.

arschles commented on pkg/controller/controller_instance.go:
```diff
@@ -1843,7 +1843,12 @@ func (c *controller) serviceInstanceHasExistingBindings(instance *v1beta1.Servic
-			return true, nil
+			for _, condition := range binding.Status.Conditions {
+				if condition.Type == v1beta1.ServiceBindingConditionReady && condition.Status != v1beta1.ConditionFalse {
+					return true, nil
```
I don't remember - what will happen when the user tries to delete the binding later? If fails b/c can't find the instance, I'd rather see that fixed in the same PR

kibbles-n-bytes replied:
@arschles If the binding has its `UnbindStatus` as `StatusNotRequired` (aka. there was never a request sent to the broker), it will be successfully deleted without having to look for the instance.

witlessbird commented:
* updated to use ```binding.Status.UnbindStatus == v1beta1.ServiceBindingUnbindStatusRequired``` to check whether a service has actual bindings.

witlessbird commented:
@kibbles-n-bytes: out of curiosity -- what would happen if a service was deleted after https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_binding.go#L215 but before https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_binding.go#L689?

kibbles-n-bytes commented:
@witlessbird Hmm... For already provisioned instances, it seems to me there'd be a race. There's no way for the instance reconciler to know if a binding request has truly been sent to the broker or not, as it might just be looking at an out-of-date binding resource in the cache.

This is a tricky situation. We should probably limit the scope of this PR to just instances with DeprovisionStatus `NotRequired` corresponding to instances either never provisioned. In this case, we know:
1) There's no race with setting the `DeletionTimestamp` of the instance vs setting `DeprovisionStatus` to `Required`, as if the API server accepts a deletion of the instance, then the resource version will update and cause [here](https://github.com/kubernetes-incubator/service-catalog/blob/789c4b27e41ca619563fa2edf60a622a286220c0/pkg/controller/controller_instance.go#L584) to fail, which will make the reconciliation retry and never communicate with the broker.
2) There's no race with deleting the instance vs sending a bind request to the broker, as the binding would never get past [here](https://github.com/kubernetes-incubator/service-catalog/blob/789c4b27e41ca619563fa2edf60a622a286220c0/pkg/controller/controller_binding.go#L315), since the instance never would have become Ready at any time.

We might also be able to include DeprovisionStatus `Succeeded`, corresponding to instances that have either been orphan mitigated or have been deprovisioned already, but I need to think a bit more about the consequences of this, especially if we want to start to retry provision failures.

witlessbird commented:
* updated to verify ```UnbindStatus``` only when ```DeprovisionStatus``` is ```ServiceInstanceDeprovisionStatusNotRequired```

witlessbird commented:
I'll check if the above change will work for DeprovisionStatus ```Succeeded``` too.

witlessbird commented:
* removed unused import

witlessbird commented:
@kibbles-n-bytes: On the second thought I'm not sure I understand your comment re: DeprovisionStatus ```Succeeded```. Wouldn't getting to this state require removal of available bindings?

kibbles-n-bytes commented:
@witlessbird Thinking about it now, it's safe; I was just heading out the office and didn't want to commit to it until I walked myself through the logic a bit more. If we have a DeprovisionStatus `Succeeded`, that implies that the instance has been orphan mitigated and at no point got a condition `Ready`, so no bindings could have gone through. And at that point, if we wanted to allow provision retries in the future, we'd be in the same situation as a DeprovisionStatus of `NotRequired`, which is safe as per above.

With that, the implementation could actually be just moving the `hasExistingBindings` check after the graceful deletion quick-exit block that currently immediately follows it, and no changes would be needed inside `c.serviceInstanceHasExistingBindings`.

witlessbird commented:
> With that, the implementation could actually be just moving the hasExistingBindings check after the graceful deletion quick-exit block that currently immediately follows it, and no changes would be needed inside c.serviceInstanceHasExistingBindings.

Done. Also added a comment why we aren't trying to filter out bindings for which binding requests have not been sent.

kibbles-n-bytes commented:
@witlessbird Thanks! You'll need to rebase now that some refactor work has gone in, but it shouldn't be too bad. Also could we get a couple unit tests for the `NotRequired` and `Succeeded` cases respectively?

witlessbird commented:
* rebased.

> Also could we get a couple unit tests for the NotRequired and Succeeded cases respectively?

It appears these are already covered in https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_instance_test.go#L1732 and  https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_instance_test.go#L1839 respectively?

kibbles-n-bytes commented:
@witlessbird Those tests are for instances without any bindings. We should add unit tests confirming that even with bindings, the instances in those cases are deleted successfully. I'd recommend not just adding a binding to those existing tests, though, as I think the cases are unique enough to warrant separate tests.

witlessbird commented:
* Added tests to verify instance deletion with bindings present.

@kibbles-n-bytes: There's now quite a bit of code duplication in these tests -- perhaps we should parametrize them?

witlessbird commented:
* signed CLA

witlessbird commented:
@kibbles-n-bytes: any chance you could take another look at this?

kibbles-n-bytes commented on pkg/controller/controller_instance_test.go:
```diff
@@ -1771,6 +1771,52 @@ func TestReconcileServiceInstanceDeleteFailedProvisionWithoutRequest(t *testing.
+// TestReconcileServiceInstanceDeleteFailedProvisionWithoutRequestWithActiveBinding tests
+// that an instance that failed to provision without making a provision request
+// will be finalized even if it has binding(s) and no deprovision request will be sent to the broker.
+func TestReconcileServiceInstanceDeleteFailedProvisionWithoutRequestWithActiveBinding(t *testing.T) {
```
"Active" binding seems to imply to me that the binding resource was created at the broker, which isn't the case here. "Inactive" seems to convey the correct intent; the k8s resource was created, but the binding it represents hasn't been created yet.

kibbles-n-bytes commented on pkg/controller/controller_instance_test.go:
```diff
@@ -1771,6 +1771,52 @@ func TestReconcileServiceInstanceDeleteFailedProvisionWithoutRequest(t *testing.
+	sharedInformers.ClusterServiceBrokers().Informer().GetStore().Add(getTestClusterServiceBroker())
+	sharedInformers.ClusterServiceClasses().Informer().GetStore().Add(getTestClusterServiceClass())
+	sharedInformers.ClusterServicePlans().Informer().GetStore().Add(getTestClusterServicePlan())
+	sharedInformers.ServiceBindings().Informer().GetStore().Add(getTestServiceBinding())
```
`getTestServiceBinding()` is a bit out of place here, because the resource it's creating is supposed to represent a binding that was created at the broker, and thus it has its `UnbindStatus` set to `Required` (looks like we don't set the Ready condition on it, though, which is a bit bizzare). Maybe we should create a new `getTestServiceBindingInactive()` function that represents a fresh binding resource from the broker.

witlessbird commented:
* rebased
* parametrized tests in ```controller_instance_test.go``` to reduce code duplication
* added a helper function to create a new, "inactive" service binding

witlessbird commented:
* fixed formatting

arschles commented:
@kibbles-n-bytes answered my question (thanks!) so I've dismissed the 'x' on my review

jboyd01 commented:
LGTM, thanks a lot @witlessbird
